icicle/backend/cpu/include/ntt_task.h

```diff
@@ -0,0 +1,1480 @@
+using namespace icicle;
+namespace ntt_cpu {
+
+  /**
+   * @brief Represents a task in the NTT computation, handling either NTT calculation or reordering.
+   *
+   * This class manages tasks within the NTT computation, performing either the NTT computation
+   * for a given sub-NTT or reordering the output if required.
+   *
+   * @method void execute() Executes the task, either performing the NTT computation or reordering the output.
+   * @method NttTaskCoordinates get_coordinates() const Returns the task's coordinates.
+   * @method void set_coordinates(NttTaskParams<S, E> params) Sets the task parameters.
+   * @method void set_data(NttData<S, E>& data) Sets the NTT data for the task.
+   */
+  template <typename S = scalar_t, typename E = scalar_t>
+  class NttTask : public TaskBase
+  {
+  public:
+    NttTask() : ntt_data(nullptr) {}
+
+    void execute();
+    NttTaskCoordinates* get_coordinates() const
+    {
+      return ntt_task_coordinates;
+    } // Returns a pointer to the `NttTaskCoordinates` structure that specifies the task's position within the NTT
+      // computation hierarchy.
+    void set_coordinates(NttTaskCoordinates* task_c_ptr)
+    {
+      ntt_task_coordinates = task_c_ptr;
+    } // Assigns a pointer to a `NttTaskCoordinates` structure, which specifies the task's position within the NTT
+      // computation hierarchy.
+    void set_data(NttData<S, E>& data)
+    {
+      ntt_data = &data;
+    } // Sets the `NttData` structure that contains all necessary data and configurations required.
+
+  private:
+    NttTaskCoordinates* ntt_task_coordinates = nullptr;
+    NttData<S, E>* ntt_data = nullptr;
+    eIcicleError reorder_and_refactor_if_needed();
+    void apply_coset_multiplication(E* current_elements, const std::vector<uint32_t>& index_in_mem, const S* twiddles);
+    eIcicleError hierarchy_0_cpu_ntt();
+    void ntt8win();
+    void ntt16win();
+    void ntt32win();
+    void hierarchy_0_dit_ntt();
+    void hierarchy_0_dif_ntt();
+    void reorder_by_bit_reverse();
+    void refactor_output_hierarchy_0();
+    uint64_t idx_in_mem(NttTaskCoordinates* ntt_task_coordinates, uint32_t element);
+  };
+
+  //////////////////////////// NttTask Implementation ////////////////////////////
+
+  /**
+   * @brief Executes the NTT task.
+   *
+   * This function determines the type of task to perform based on the `reorder` flag in `ntt_task_coordinates`.
+   * - If `reorder` is `false` (most tasks), it performs the NTT computation by invoking `hierarchy_0_cpu_ntt()`.
+   * - If `reorder` is `true`, it handles reordering and refactoring by calling `reorder_and_refactor_if_needed()`.
+   * @return void
+   */
+  template <typename S, typename E>
+  void NttTask<S, E>::execute()
+  {
+    if (!ntt_task_coordinates->reorder) {
+      hierarchy_0_cpu_ntt();
+    } else {
+      // if all hierarchy_0_subntts are done, and at least 2 layers in hierarchy 0 - reorder the subntt's output
+      reorder_and_refactor_if_needed();
```

Comment:
can't we do that on the same task?
i.e. reorder and then continue with the same task to execute the ntt.